### ISSUE TITLE
mypy 1.8.0 regression: Incorrectly flags unreachable with re.match() result

### ISSUE THREAD

**Issue by amykyta3** [python/mypy#16770]
**Bug Report**

It seems that v1.8.0 regressed in how it handles type checking with returned match objects from Python's regex stdlib.
It now flags unreachable code when checking on the truthyness of a regex match return value.

This was not the case in v1.7.x

**To Reproduce**

```python
import re
import sys

if re.match(r"\d*", sys.argv[1]):
    print("a")
else:
    print("b") # mypy 1.8.0 incorrectly flags as unreachable
```

Results in:
>tst.py:7: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 1 source file)

Changing the line to explicitly cast it as a boolean avoids this:
```python
if bool(re.match(r"\d*", sys.argv[1])):
```
or:
```python
if re.match(r"\d*", sys.argv[1]) is not None:
```

This seems to be related to the config setting of `strict_optional = False`

**Expected Behavior**

I would not expect any unreachable lines in this example.

**Actual Behavior**

mypy 1.8.0 reports an unreachable violation.

**Your Environment**

- Mypy version used: 1.8.0
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files):
```
[mypy]
warn_unreachable = True
strict_optional = False
```

- Python version used: Python 3.10.12

**Comment by hauntsaninja** [comment]
Duplicate of https://github.com/python/mypy/issues/16718
